### PR TITLE
This is mainly a pull request to fix tlRender to be used as a git submodule

### PR DESCRIPTION
**Pull request by ggarra13**
Mainly, it changes the CMAKE_SOURCE_DIR to CMAKE_CURRENT_SOURCE dir.  There's also the addition of play-fltk but I removed it at the end.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -118,7 +119,8 @@ endif()
 if(TLRENDER_BUILD_GL)
     find_package(OpenGL REQUIRED)
     if(TLRENDER_BUILD_PROGRAMS OR TLRENDER_BUILD_EXAMPLES)
-        find_package(GLFW REQUIRED)
+	find_package(GLFW REQUIRED)
+	find_package(FLTK NO_MODULE )
```
Comment (darbyjohnston) [inline CMakeLists.txt:123]: Is this still necessary if the ```play-fltk``` example was removed?

File: README.md [hunk 2]
```diff
@@ -54,6 +54,7 @@ Optional dependencies:
 * [OpenEXR](https://www.openexr.com/)
 * [PNG](https://libpng.sourceforge.io/index.html)
 * [TIFF](http://www.libtiff.org)
+* [FLTK](https://www.fltk.org)
```
Comment (darbyjohnston) [inline README.md:57]: Is this still necessary if the play-fltk example was removed?

File: etc/SuperBuild/cmake/Modules/BuildOpenEXR.cmake [hunk 3]
```diff
@@ -15,8 +15,8 @@ ExternalProject_Add(
     DEPENDS Imath ZLIB
     GIT_REPOSITORY ${OpenEXR_GIT_REPOSITORY}
     GIT_TAG ${OpenEXR_GIT_TAG}
-    PATCH_COMMAND ${CMAKE_COMMAND} -E copy
-        ${CMAKE_SOURCE_DIR}/OpenEXR-patch/src/lib/OpenEXR/ImfDwaCompressor.cpp
-        ${CMAKE_CURRENT_BINARY_DIR}/OpenEXR/src/OpenEXR/src/lib/OpenEXR/ImfDwaCompressor.cpp
+    PATCH_COMMAND ${CMAKE_COMMAND} -E copy_if_different
```
Comment (darbyjohnston) [inline etc/SuperBuild/cmake/Modules/BuildOpenEXR.cmake:18]: Is this was necessary it should probably be changed in all of the build files.

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -1,6 +1,6 @@
 add_subdirectory(test-patterns)
 if(TLRENDER_BUILD_GL)
-    add_subdirectory(play-glfw)
+  add_subdirectory(play-glfw)
```
Comment (darbyjohnston) [inline examples/CMakeLists.txt:3]: It's generally better to make formatting changes separately so it's easier to review the functional changes. But if the tab is removed it should be replaced with 4 spaces. I've accidentally added a bunch of tabs and should probably do a global search and replace.

**Review by darbyjohnston — COMMENTED** [review]
Thanks for this, I'm definitely interested in having tlRender work as a git submodule. Is it worthwhile to add a small example to the README?

**Comment by ggarra13** [comment]
> El 10 sep. 2022, a las 17:57, Darby Johnston ***@***.***> escribió:
> 
> 
> @darbyjohnston commented on this pull request.
> 
> Thanks for this, I'm definitely interested in having tlRender work as a git submodule. Is it worthwhile to add a small example to the README?
> 

Not sure what example, as it takes a tad to setup.

> In CMakeLists.txt:
> 
> > @@ -118,7 +119,8 @@ endif()
>  if(TLRENDER_BUILD_GL)
>      find_package(OpenGL REQUIRED)
>      if(TLRENDER_BUILD_PROGRAMS OR TLRENDER_BUILD_EXAMPLES)
> -        find_package(GLFW REQUIRED)
> +	find_package(GLFW REQUIRED)
> +	find_package(FLTK NO_MODULE )
> 
> Is this still necessary if the play-fltk example was removed?

No, you are right. Remove all references to FLTK.

> 
> In README.md:
> 
> > @@ -54,6 +54,7 @@ Optional dependencies:
>  * [OpenEXR](https://www.openexr.com/)
>  * [PNG](https://libpng.sourceforge.io/index.html)
>  * [TIFF](http://www.libtiff.org)
> +* [FLTK](https://www.fltk.org)
> 
> Is this still necessary if the play-fltk example was removed?
> 

No, again.

> In etc/SuperBuild/cmake/Modules/BuildOpenEXR.cmake:
> 
> > @@ -15,8 +15,8 @@ ExternalProject_Add(
>      DEPENDS Imath ZLIB
>      GIT_REPOSITORY ${OpenEXR_GIT_REPOSITORY}
>      GIT_TAG ${OpenEXR_GIT_TAG}
> -    PATCH_COMMAND ${CMAKE_COMMAND} -E copy
> -        ${CMAKE_SOURCE_DIR}/OpenEXR-patch/src/lib/OpenEXR/ImfDwaCompressor.cpp
> -        ${CMAKE_CURRENT_BINARY_DIR}/OpenEXR/src/OpenEXR/src/lib/OpenEXR/ImfDwaCompressor.cpp
> +    PATCH_COMMAND ${CMAKE_COMMAND} -E copy_if_different
> 
> Is this was necessary it should probably be changed in all of the build files.
> 

I didn’t find anywhere else where you were using copy of a file, but if you are then, yes, changing it to copy_if_different is better as it speeds up the compilation as the dependency is not regenerated.

> In examples/CMakeLists.txt:
> 
> > @@ -1,6 +1,6 @@
>  add_subdirectory(test-patterns)
>  if(TLRENDER_BUILD_GL)
> -    add_subdirectory(play-glfw)
> +  add_subdirectory(play-glfw)
> 
> It's generally better to make formatting changes separately so it's easier to review the functional changes. But if the tab is removed it should be replaced with 4 spaces. I've accidentally added a bunch of tabs and should probably do a global search and replace.

My bad too.  I had accidentally left emacs c/c++ mode use tabs in favor of spaces.  I modified my .emacs file and will copy it to all my machines.

—
Gonzalo Garramuno
***@***.***

**Comment by darbyjohnston** [comment]
No worries about the git submodule example, it was just a suggestion if it was a simple thing to do. Just checking-in to see if you can make the other changes and push them to this PR? Thanks!

**Comment by ggarra13** [comment]
On 18/9/22 15:57, Darby Johnston wrote:
>
> No worries about the git submodule example, it was just a suggestion 
> if it was a simple thing to do. Just checking-in to see if you can 
> make the other changes and push them to this PR? Thanks!
>
I haven't made changes to tlRender as I've been too busy with 
implementing the features in mrViewer.  I am mostly done so I'll start 
tweaking things in tlRender now (hopefully I wont break things).

The first thing I want to do is to get a member function in 
tl::gl::Render to retrieve the shaders.  I want to draw text with NDC 
coordinates for the HUD so that they always stay in the same size and in 
the same position, instead of how it is now that they have the same 
matrix than the image.

Once I am done with that, I'll send another PR request.

-- 
Gonzalo Garramuño
***@***.***

**Comment by darbyjohnston** [comment]
It would be better to keep the PRs separate so it's easier to review the changes. If this PR is for fixing tlRender to be used as a git submodule it should just contain the CMake changes. If there are other changes you are interested in making they should be separate PRs.

Also I think we need to find a different approach for the text besides getting the shader source code. Getting the source code breaks the encapsulation for the renderer, it's really supposed to be private. The renderer should ideally be independent of the underlying graphics API, one of my next tasks is writing Metal and Direct X versions.

**Comment by ggarra13** [comment]
> El 18 sep. 2022, a las 20:29, Gonzalo Garramuño ***@***.***> escribió:

> I haven't made changes to tlRender as I've been too busy with implementing the features in mrViewer.  I am mostly done so I'll start tweaking things in tlRender now (hopefully I wont break things).
> 
> The first thing I want to do is to get a member function in tl::gl::Render to retrieve the shaders.  I want to draw text with NDC coordinates for the HUD so that they always stay in the same size and in the same position, instead of how it is now that they have the same matrix than the image.
> 
> Once I am done with that, I'll send another PR request.
> 

Find the new PR as tlRender-gga/my-feature.  I am still getting familiar with how to use GitHub and git with multiple users (I apologize for the previous PR which had a lot of things mixed in).

—
Gonzalo Garramuno
***@***.***

**Comment by darbyjohnston** [comment]
No worries, I'm happy to answer any questions about it. Once you get used to working with multiple PRs on GitHub it makes contributing a fairly seamless process.

**Comment by ggarra13** [comment]
> El 18 sep. 2022, a las 21:14, Darby Johnston ***@***.***> escribió:
> 
> Also I think we need to find a different approach for the text besides getting the shader source code. Getting the source code breaks the encapsulation for the renderer, it's really supposed to be private. The renderer should ideally be independent of the underlying graphics API, one of my next tasks is writing Metal and Direct X versions.
> 

Oh.  I just saw this.  I am not getting the source code, just the shader itself to set the uniform of the mvp matrix.

I just need to get access to the matrix of the text shader so I can draw text in raster space for the HUD.  Currently, I am doing the following (after the patch in the PR to get the shader):


        const auto viewportSize = _getViewportSize();
        const glm::mat4x4 vpm = glm::ortho(
            0.F,
            static_cast<float>(viewportSize.w),
            static_cast<float>(viewportSize.h),
            0.F,
            1.F,
            -1.F);

        auto mvp = math::Matrix4x4f(
            vpm[0][0], vpm[0][1], vpm[0][2], vpm[0][3],
            vpm[1][0], vpm[1][1], vpm[1][2], vpm[1][3],
            vpm[2][0], vpm[2][1], vpm[2][2], vpm[2][3],
            vpm[3][0], vpm[3][1], vpm[3][2], vpm[3][3] );

        auto shader = gl.render->getShader( "text" );
        shader->bind();
        shader->setUniform("transform.mvp", mvp);

        if ( p.hud & HudDisplay::kDirectory )
            _drawText( p.fontSystem->getGlyphs(directory, fontInfo), pos,
                       lineHeight, labelColor );

As you can see, the change is in tlGL/Render.cpp and in my own GL class.  That should not prevent it from working with Metal, I think.

—
Gonzalo Garramuno
***@***.***

**Comment by ggarra13** [comment]
I am closing this pull request as I opened two others: cmake changes and getShader.

**Comment by ggarra13** [comment]
> El 18 sep. 2022, a las 21:14, Darby Johnston ***@***.***> escribió:
> 
> 
> Also I think we need to find a different approach for the text besides getting the shader source code. Getting the source code breaks the encapsulation for the renderer, it's really supposed to be private. The renderer should ideally be independent of the underlying graphics API, one of my next tasks is writing Metal and Direct X versions.
> 

But currently, you have GLSL shader source code in tlQtWidget/TimelineViewport.cpp.

Also, I looked briefly into how to pass Uniform data in DirectX shaders and it requires to pass the projection matrix as XMFLOAT4 instead of standard floats to keep the alignment consistent. 

https://learn.microsoft.com/en-us/windows/uwp/gaming/porting-uniforms-and-attributes



—
Gonzalo Garramuno
***@***.***

**Pull request closed without merging**